src/System.Memory/src/System/Span.cs

```diff
@@ -323,37 +322,87 @@ public void CopyTo(Span<T> destination)
-                IntPtr srcMinusDst = Unsafe.ByteOffset<T>(ref dst, ref src);
-                int length = _length;
 
-                bool srcGreaterThanDst = (sizeof(IntPtr) == sizeof(int)) ? srcMinusDst.ToInt32() >= 0 : srcMinusDst.ToInt64() >= 0;
-                if (srcGreaterThanDst)
+                if (!TryFastCopy(ref dst, destLength, ref src, length))
                 {
-                    // Source address greater than or equal to destination address. Can do normal copy.
-                    for (int i = 0; i < length; i++)
+                    IntPtr srcMinusDst = Unsafe.ByteOffset<T>(ref dst, ref src);
+                    bool srcGreaterThanDst = (sizeof(IntPtr) == sizeof(int)) ? srcMinusDst.ToInt32() >= 0 : srcMinusDst.ToInt64() >= 0;
+                    if (srcGreaterThanDst)
                     {
-                        Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        // Source address greater than or equal to destination address. Can do normal copy.
+                        for (int i = 0; i < length; i++)
+                        {
+                            Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        }
                     }
-                }
-                else
-                {
-                    // Source address less than destination address. Must do backward copy.
-                    int i = length;
-                    while (i-- != 0)
+                    else
                     {
-                        Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        // Source address less than destination address. Must do backward copy.
+                        int i = length;
+                        while (i-- != 0)
+                        {
+                            Unsafe.Add<T>(ref dst, i) = Unsafe.Add<T>(ref src, i);
+                        }
                     }
                 }
+
+                return true;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe bool TryFastCopy(ref T dst, int dstLength, ref T src, int srcLength)
+        {
+            IntPtr srcMinusDst = Unsafe.ByteOffset<T>(ref dst, ref src);
+            bool srcGreaterThanDst = (sizeof(IntPtr) == sizeof(int)) ? srcMinusDst.ToInt32() >= 0 : srcMinusDst.ToInt64() >= 0;
+            IntPtr tailDiff;
+
+            if (srcGreaterThanDst)
+            {
+                tailDiff = Unsafe.ByteOffset<T>(ref Unsafe.Add<T>(ref dst, dstLength), ref src);
+            }
+            else
+            {
+                tailDiff = Unsafe.ByteOffset<T>(ref Unsafe.Add<T>(ref src, srcLength), ref dst);
+            }
+
+            bool isOverlapped = (sizeof(IntPtr) == sizeof(int)) ? tailDiff.ToInt32() < 0 : tailDiff.ToInt64() < 0;
+            if (!isOverlapped && !SpanHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                CopyBlocks(ref dst, ref src, srcLength);
```

Comment:
> Why can't we call CopyBlocks if SpanHelpers.IsReferenceOrContainsReferences

Because of CopyBlocks does not have the correct GC tracking (does not update cards for generational GC). 

If you would like to see the details, take a look at the optimized Copy implementation in CoreCLR that updates the cards in bulk (https://github.com/dotnet/coreclr/pull/9999).